### PR TITLE
DOCKER-728: read/write "triton.*" labels from/to VM.tags["triton.*"]

### PR DESCRIPTION
**Pull request by trentm**
DOCKER-721: Docker 1.10 ps/inspect errors out when label contains boolean values

### CODE REVIEW

File: lib/backends/sdc/containers.js [hunk 1]
```diff
@@ -1475,6 +1472,9 @@ function buildVmPayload(opts, container, callback) {
     };
     var restartPolicy;
     var volumesFrom;
+    var i;
+    var labelKeys;
+    var labelKey;
```
Comment (joshwilsdon) [inline lib/backends/sdc/containers.js]: These aren't in the correct place alphabetically.

File: lib/backends/sdc/containers.js [hunk 2]
```diff
@@ -1507,15 +1507,28 @@ function buildVmPayload(opts, container, callback) {
...
+                continue;
             }
-            payload.tags[common.LABELTAG_PREFIX + key] = container.Labels[key];
-        });
+            if (triton_tags.isTritonTag(labelKey)) {
+                try {
+                    payload.tags[labelKey] = triton_tags.validateTritonTagStr(
+                        labelKey, container.Labels[labelKey]);
+                } catch (err) {
+                    callback(err);
```
Comment (joshwilsdon) [inline lib/backends/sdc/containers.js]: Would it be good to turn this into a more specific error so we don't end up with a generic one to the user?

File: lib/backends/sdc/utils.js [hunk 3]
```diff
@@ -5,32 +5,71 @@
...
+ * Note: Docker label values must be strings otherwise the docker 1.10 client
+ * will blow up. See https://github.com/docker/docker/issues/20881 and
+ * DOCKER-721. Currently VM.tags are limited to string, boolean or number,
+ * so we can simply `.toString()`.
+ */
 function dockerLabelsFromVmTags(tags) {
     var labels = {};
-    var labelPrefix = common.LABELTAG_PREFIX;
+    const LABELTAG_PREFIX = common.LABELTAG_PREFIX;
+    const expectedTypes = {string: true, boolean: true, number: true};
```
Comment (joshwilsdon) [inline lib/backends/sdc/utils.js]: should these be alphabetical?

File: lib/backends/sdc/containers.js (hunk 1 above)
Comment (trentm) [inline lib/backends/sdc/containers.js]: Ah didn't know we were doing alpha. Will update.

File: lib/backends/sdc/containers.js (hunk 2 above)
Comment (trentm) [inline lib/backends/sdc/containers.js]: Yes definitely. Thanks. I'd forgotten to do that.

FWIW, I left triton-tags.js generic (i.e. not using the local errors.js) so that it could eventually be pulled out to a separate module (node-triton-tags.js or whatever) to have the validation shared between here and in VMAPI.

File: lib/backends/sdc/utils.js (hunk 3 above)
Comment (trentm) [inline lib/backends/sdc/utils.js]: will do

Comment (joshwilsdon) [inline lib/backends/sdc/utils.js]: ooc: why object for expectedTypes instead of array?

File: lib/triton-tags.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+ * For now, using the more limited labels allowed by RFC1123. RFC2181 supercedes
+ * 1123, but the broader range of characters can sometimes cause problems with
+ * other systems (e.g. see the underscore in RFC5321).
+ */
+var DNS_NAME_RE = /^[a-z0-9][a-z0-9\-]{0,62}(?:\.[a-z0-9][a-z0-9\-]{0,62})*$/i;
+
+const TRITON_TAG_PREFIX = 'triton.';
+
+var typeFromKey = {
+    'triton.cns.services': 'string',
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: should these be:

```
TRITON_TAG_PREFIX + '.cns.services': 'string',
```

?

File: lib/triton-tags.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    'triton.cns.reverse_ptr': 'string'
+};
+
+/*
+ * Validator functions take a `val` (already checked to be of the appropriate
+ * type per `typeFromKey`) and should throw an `Error` if invalid. The don't
+ * return a value.
+ */
+var validatorFromKey = {
+    'triton.cns.services': function validateTritonCnsServices(val) {
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: TRITON_TAG_PREFIX + '.cns.services'?

File: lib/triton-tags.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        for (var i = 0; i < fqdns.length; i++) {
+            var fqdn = fqdns[i];
+            if (fqdn.length > 255 || !fqdn.match(DNS_NAME_RE)) {
+                unsafes.push(fqdn);
+            }
+        }
+
+        if (unsafes.length > 0) {
+            throw new Error(format(
+                'invalid "triton.cns.services" tag: "%s" %s not DNS safe',
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: TRITON_TAG_PREFIX?

File: lib/triton-tags.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+var typeFromKey = {
+    'triton.cns.services': 'string',
+    'triton.cns.disable': 'boolean',
+    'triton.cns.reverse_ptr': 'string'
+};
+
+/*
+ * Validator functions take a `val` (already checked to be of the appropriate
+ * type per `typeFromKey`) and should throw an `Error` if invalid. The don't
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: *They don't

File: lib/backends/sdc/containers.js [hunk 8]
```diff
@@ -1507,15 +1507,28 @@ function buildVmPayload(opts, container, callback) {
...
                 // PACKAGE_SELECTION_LABEL is a special label. It gets consumed
                 // here and doesn't end up in the user's list. Instead it's used
                 // in getPackage().
-                return;
+                continue;
             }
-            payload.tags[common.LABELTAG_PREFIX + key] = container.Labels[key];
-        });
+            if (triton_tags.isTritonTag(labelKey)) {
+                try {
```
Comment (joshwilsdon) [inline lib/backends/sdc/containers.js]: By having these validateTritonTag* functions throw and catching in this `try {`, we lose the ability to distinguish between programmer errors and runtime errors. For example we've validated that we have a triton tag here, but later we'll do:

```
assert.func(validator, 'validator for tag ' + key);
```

or:

```
if (type === undefined) {
    throw new Error('Unrecognized special triton tag "' + key + '"');
}
```

which seem like they would only happen on programmer error. But they'll be caught in this `try {` rather than throwing an uncaught exception and generating a core. Is this fine? It might be, just wanted to point it out.

File: lib/triton-tags.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        }
+    }
+};
+
+
+// --- exports
+
+/*
+ * Return true if the given key uses the Triton tag prefix.
+ * Note that it still might not be one of the specific defined tag
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: *tag**s**

File: lib/backends/sdc/utils.js (hunk 3 above)
Comment (trentm) [inline lib/backends/sdc/utils.js]: easier lookup that arr.indexOf, basically. I don't have a string pref.
If JS had Python's `in` operator, I'd use an array for this kind of thing much more frequently.

File: lib/triton-tags.js (hunk 4 above)
Comment (trentm) [inline lib/triton-tags.js]: I suppose, yes. Though I'd be a little sad. I expect more grepping for "triton.cns.services" than for "cns.services".

File: lib/triton-tags.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    assert.func(validator, 'validator for tag ' + key);
+    validator(val);
+
+    return val;
+}
+
+/*
+ * Convert and validate the given Triton tag string value.
+ *
+ * @returns {String} On success, the validated value is returned.
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: This seems incorrect in that the return will not always be a String. It will be one of String/Boolean/Number, correct?

File: lib/triton-tags.js [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+    var validator = validatorFromKey[key];
+    assert.func(validator, 'validator for tag ' + key);
+    validator(val);
+
+    return val;
+}
+
+/*
+ * Convert and validate the given Triton tag string value.
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: Might be good to be a bit more specific about the conversion intended here.

File: lib/triton-tags.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        } else if (str === 'false') {
+            val = false;
+        } else {
+            throw new Error(format(
+                'Triton tag "%s" value must be "true" or "false": %j',
+                key, str));
+        }
+        break;
+    case 'number':
+        val = Number(str);
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: Do we care about an empty string here? Number('') will be 0.

File: lib/triton-tags.js [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    // Validate.
+    var validator = validatorFromKey[key];
+    assert.func(validator, 'validator for tag ' + key);
+    validator(val);
+
+    return val;
+}
+
+
+module.exports = {
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: Should this also export TRITON_TAG_PREFIX in case callers (like cli-triton-tags.test.js) want to use the prefix, so they don't also need to hardcode this value?

File: lib/backends/sdc/containers.js (hunk 8 above)
Comment (trentm) [inline lib/backends/sdc/containers.js]: Exploring that, which do you like best:

```
// Returns value, or throws Error.
try {
    payload.tags[labelKey] = triton_tags.validateTritonTagStr(
        labelKey, container.Labels[labelKey]);
} catch (err) {
    callback(new errors.ValidationError(err, err.message));
    return;
}

// Returns one value: the validated value or an Error instance.
var errOrVal = triton_tags.validateTritonTagStr(
    labelKey, container.Labels[labelKey]);
if (errOrVal instanceof Error) {
    callback(new errors.ValidationError(errOrVal,
        errOrVal.message));
    return;
} else {
    payload.tags[labelKey] = errOrVal;
}

// Returns a tuple `[err, val]`.
var rv = triton_tags.validateTritonTagStr(   // [err, val]
    labelKey, container.Labels[labelKey]);
if (rv[0]) {
    callback(new errors.ValidationError(rv[0], rv[0].message));
    return;
} else {
    payload.tags[labelKey] = rv[1];
}
```

?

File: lib/triton-tags.js (hunk 10 above)
Comment (trentm) [inline lib/triton-tags.js]: You are correct. Updated to:

```
 * @returns {String|Boolean|Number} On success, the validated value is returned.
```

File: lib/triton-tags.js (hunk 11 above)
Comment (trentm) [inline lib/triton-tags.js]: Perhaps call it `parseTritonTagStr` (and leave the other one as `validateTritonTag` -- that one intended to be used by VMAPI which gets tags with the proper type)?

File: lib/triton-tags.js (hunk 12 above)
Comment (trentm) [inline lib/triton-tags.js]: I also forgot to give isNaN an arg there. Corrected:
```
        if (isNaN(val)) {
```

File: lib/triton-tags.js [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+ * return a value.
+ */
+var validatorFromKey = {
+    'triton.cns.services': function validateTritonCnsServices(val) {
+        assert.string(val, 'val');
+
+        var unsafes = [];
+        var fqdns = val.split(',');
+        for (var i = 0; i < fqdns.length; i++) {
+            var fqdn = fqdns[i];
```
Comment (richardkiene) [inline lib/triton-tags.js]: This logic looks like it came from VMAPI and isn't quite correct. I have a change I'll be merging in (thanks to @arekinath) to VMAPI shortly and we'll want that same logic here if it must be copied. The new change allows for SRV records to be used with CNS (e.g. triton.cns.services=jabber:5223).

Should we think about a way to better re-use code?

File: lib/triton-tags.js (hunk 12 above)
Comment (trentm) [inline lib/triton-tags.js]: Thanks. I'll get it to disallow empty string.

File: lib/backends/sdc/containers.js (hunk 8 above)
Comment (joshwilsdon) [inline lib/backends/sdc/containers.js]: I'd propose 2 more options, first the one I'd probably use since I believe it's the idiomatic node way to do this:

```
// Calls callback(err, value)
triton_tags.validateAndNormalizeTritonTagStr(labelKey, container.Labels[labelKey], function _validateCb(err, value) {
    if (err) {
        callback(err);
        return;
    }

    payload.tags[labelKey] = value;
});
```

or if the reason for it being invalid isn't important (like the current code):

```
// separate validation and normalization
if (triton_tags.validTritonTagStr(labelKey, container.Labels[labelKey])) {
    payload.tags[labelKey] = triton_tags.normalizeTritonTagStr(labelKey, container.Labels[labelKey]));
} else {
    callback(new errors.ValidationError('invalid ' + labelKey));
    return;
}
```

File: test/integration/cli-triton-tags.test.js [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+        cli.run(t, {
+            args: args,
+            /* JSSTYLED */
+            expectedErr: 'Error response from daemon: Unrecognized special triton tag "triton.bogus"'
+        });
+    });
+});
+
+
+test('teardown', cli.rmAllCreated);
```
Comment (joshwilsdon) [inline test/integration/cli-triton-tags.test.js]: We seem to have tested boolean and string values, should we also add a test for a triton.cns tag that has a numeric value so we test the:

```
case 'number':
```

codepath? With both a valid and invalid number and maybe a float and an empty string?

Comment (trentm) [inline test/integration/cli-triton-tags.test.js]: @joshwilsdon There aren't any current `triton.` tags that take a number.

File: lib/triton-tags.js (hunk 13 above)
Comment (trentm) [inline lib/triton-tags.js]: Is there a case where it is needed? I actually intentionally didn't export it: users of the module should use `isTritonTag(key)` instead.

File: lib/triton-tags.js (hunk 4 above)
Comment (trentm) [inline lib/triton-tags.js]: I'm inclined to not do this step, because JS doesn't allow an expression for the key part of an object literal:

```
var typeFromKey = {
    TRITON_TAG_PREFIX + 'cns.services': 'string',   // <--- syntax error
...
```

I wonder if an acceptable alternative (if really wanted) is a top-level assert that all the keys in the
`typeFromKey` and `validatorFromKey` start with the prefix?

File: test/lib/vm.js [hunk 16]
```diff
@@ -51,6 +51,104 @@ function getVm(t, opts, callback) {
...
+            var keys = Object.keys(opts.tags);
+            for (var i = 0; i < keys.length; i++) {
+                var k = keys[i];
+                if (obj.tags[k] !== opts.tags[k]) {
+                    incomplete = true;
+                    break;
+                }
+            }
+
+            if (!incomplete) {
```
Comment (joshwilsdon) [inline test/lib/vm.js]: Since this doesn't check the job (which might have failed) this appears it could continue forever. Perhaps either add a check on the job 'execution' field or at least a timeout value such that if this runs say some hundreds of loops we'll give up?

File: lib/triton-tags.js (hunk 4 above)
Comment (joshwilsdon) [inline lib/triton-tags.js]: That's fine. I'd suggest taking out the TRITON_TAG_PREFIX then and just using 'triton.' everywhere. So that someone doesn't think they can actually change this meaningfully in one place. Or add a comment saying that this is only a convenience thing and in order to change one would still need to replace all instances of `triton.` everywhere.

File: lib/triton-tags.js (hunk 11 above)
Comment (joshwilsdon) [inline lib/triton-tags.js]: yeah, `parseTritonTagStr` I think fits better.

File: test/integration/cli-triton-tags.test.js (hunk 15 above)
Comment (joshwilsdon) [inline test/integration/cli-triton-tags.test.js]: Perhaps we shouldn't include the code to support that yet then? Since it can't be tested. It seems better for validateTritonTagStr to just not support numbers yet than to have code there that hasn't and can't be tested.

File: lib/triton-tags.js (hunk 13 above)
Comment (joshwilsdon) [inline lib/triton-tags.js]: The only reason would be if you want to be able to change it. But as I said above, that's the only good reason I see for having a constant at all. Since we're not going to make it possible to change, there's no good reason to export it. Everyone can just hardcode 'triton.' and be happy.

File: lib/triton-tags.js [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+        cb(new Error(errmsg));
+    } else {
+        cb(null, val);
+    }
+}
+
+
+module.exports = {
+    isTritonTag: isTritonTag,
+    validateTritonTag: validateTritonTag,
```
Comment (joshwilsdon) [inline lib/triton-tags.js]: Where is validateTritonTag() actually used? I must be missing something. Why is this needed?

Comment (trentm) [inline lib/triton-tags.js]: It isn't yet. It is the intended function to be used by VMAPI later today when I break this
out to a separate node-triton-tags module and have docker and vmapi share that.
However, I'd like to get this in sdc-docker first.

**Comment by joshwilsdon** [comment]
Aside from perhaps some tests for the validateTritonTag() function (or removing from the API until such time as it's needed), I'd say this lgtm.

**Comment by trentm** [comment]
Squashed and merged to master in commit 5e655d4
